ooniprobe/Model/NetworkMeasurement.mm

```diff
@@ -652,22 +690,27 @@ -(void)on_entry:(const char*)str{
         NSData *data = [[NSString stringWithUTF8String:str] dataUsingEncoding:NSUTF8StringEncoding];
         NSDictionary *json = [NSJSONSerialization JSONObjectWithData:data options:0 error:&error];
         int blocking = ANOMALY_GREEN;
+        if (error != nil) {
+#ifdef DEBUG
+            NSLog(@"Error parsing JSON: %@", error);
+#endif
+            blocking = ANOMALY_ORANGE;
+            [self updateAnomaly:blocking];
+            return;
+        }
         NSArray *keys = [[NSArray alloc] initWithObjects:@"facebook_tcp_blocking", @"facebook_dns_blocking", nil];
         for (NSString *key in keys) {
-            if ([json objectForKey:key]){
-                if ([json objectForKey:key] == [NSNull null]) {
+            if ([[json objectForKey:@"test_keys"] objectForKey:key]){
+                if ([[json objectForKey:@"test_keys"] objectForKey:key] == [NSNull null]) {
                     if (blocking < ANOMALY_ORANGE)
                         blocking = ANOMALY_ORANGE;
                 }
-                else if ([[json objectForKey:key] boolValue]) {
+                else if ([[[json objectForKey:@"test_keys"] objectForKey:key] boolValue]) {
                     blocking = ANOMALY_RED;
                 }
             }
         }
-        if (blocking > self.anomaly){
-            self.anomaly = blocking;
-            [TestStorage set_anomaly:self.test_id :blocking];
-        }
+        [self updateAnomaly:blocking];
```

Comment:
As I mentioned on slack.

I think this whole file can do with some easy refactoring, that would improve readability and maybe also prevent errors like this one from emerging in the future.

I will paste for posterity my suggestions in here:

There is a fair amount of duplication in the above code. I was suggesting to refactor the block like:

````
        int blocking = ANOMALY_GREEN;
        NSArray *keys = [[NSArray alloc] initWithObjects:@"facebook_tcp_blocking", @"facebook_dns_blocking", nil];
        for (NSString *key in keys) {
            if ([json objectForKey:key]){
                if ([json objectForKey:key] == [NSNull null]) {
                    if (blocking < ANOMALY_ORANGE)
                        blocking = ANOMALY_ORANGE;
                }
                else if ([[json objectForKey:key] boolValue]) {
                    blocking = ANOMALY_RED;
                }
            }
        }
```

Into some function that takes arguments like so:

```
func isOrangeAnomaly(testKeys, targetKeys)
```

There are also many more bits of code that can probably be easily refactored into a function and can reduce the amount of duplication.

Reply:
this only for IM tests?

Reply:
There are very similar blocks of code in the other tests as well. I would expect that with this refactoring you can reduce the amount of overall code in this file significantly.

Reply:
The fact is that this class is working well and will be totally deleted in the 2.0 version. 
Refactoring the code means that everything should be tested again and can produce bugs.
I'm against that.